src/test/kotlin/com/shankhadeepghoshal/kotlinfinaltask/service/DogsApiServiceImplTest.kt

```diff
@@ -0,0 +1,82 @@
+package com.shankhadeepghoshal.kotlinfinaltask.service
+
+import com.shankhadeepghoshal.kotlinfinaltask.exception.DogNotFoundException
+import com.shankhadeepghoshal.kotlinfinaltask.pojo.Dog
+import com.shankhadeepghoshal.kotlinfinaltask.pojo.DogApiResponseGetImages
+import com.shankhadeepghoshal.kotlinfinaltask.repository.ICustomDogRepository
+import com.shankhadeepghoshal.kotlinfinaltask.repository.IDogRepository
+import io.mockk.MockKAnnotations
+import io.mockk.coEvery
+import io.mockk.impl.annotations.InjectMockKs
+import io.mockk.impl.annotations.MockK
+import io.mockk.junit5.MockKExtension
+import kotlinx.coroutines.flow.flowOf
+import kotlinx.coroutines.flow.last
+import kotlinx.coroutines.runBlocking
+import org.junit.jupiter.api.Assertions
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.extension.ExtendWith
+
+@ExtendWith(MockKExtension::class)
+@MockKExtension.KeepMocks
+internal class DogsApiServiceImplTest {
+    private val URLS = listOf("http://www.foo.com/labrador")
+
+    @MockK
+    lateinit var customDogRepo: ICustomDogRepository
+
+    @MockK
+    lateinit var crudDogRepo: IDogRepository
+
+    @MockK
+    lateinit var restService: RestService
+
+    @InjectMockKs
+    lateinit var apiServiceImpl: DogsApiServiceImpl
+
+    @BeforeEach
+    fun setUp() {
+        MockKAnnotations.init(this)
+    }
+
+    @Test
+    fun `test find all dogs`() {
+        val dogOfInterest = Dog(1, "Labrador", "Retriever")
+        coEvery { crudDogRepo.findAll() } returns flowOf(dogOfInterest)
+        val dogReturned = runBlocking {
```

Comment:
instead of using runBlocking you can use a more convenient way:
```
 fun `test find all dogs`() = runTest {
```

`runTest requires that, after the test coroutine has completed, all the other coroutines launched inside testBody also complete, or are cancelled. Otherwise, the test will be failed (which, on JVM and Native, means that runTest itself will throw AssertionError, whereas on JS, the Promise will fail with it).`